includes/class-wc-stripe-blocks-support.php

```diff
@@ -0,0 +1,345 @@
+	 * Determine if store allows Payment Request buttons - e.g. Apple Pay / Chrome Pay.
+	 *
+	 * @return bool True if merchant has opted into payment request.
+	 */
+	private function get_allow_payment_request() {
+		$option = isset( $this->settings['payment_request'] ) ? $this->settings['payment_request'] : false;
+		return filter_var( $option, FILTER_VALIDATE_BOOLEAN );
+	}
+
+	/**
+	 * Return the button type for the payment button.
+	 *
+	 * @return string Defaults to 'default'.
+	 */
+	private function get_button_type() {
+		return isset( $this->settings['payment_request_button_type'] ) ? $this->settings['payment_request_button_type'] : 'default';
+	}
+
+	/**
+	 * Return the theme to use for the payment button.
+	 *
+	 * @return string Defaults to 'dark'.
+	 */
+	private function get_button_theme() {
+		return isset( $this->settings['payment_request_button_theme'] ) ? $this->settings['payment_request_button_theme'] : 'dark';
+	}
+
+	/**
+	 * Return the height for the payment button.
+	 *
+	 * @return string A pixel value for the height (defaults to '64').
+	 */
+	private function get_button_height() {
+		return isset( $this->settings['payment_request_button_height'] ) ? str_replace( 'px', '', $this->settings['payment_request_button_height'] ) : '64';
+	}
+
+	/**
+	 * Return the inline cc option.
+	 *
+	 * @return boolean True if the inline CC form option is enabled.
+	 */
+	private function get_inline_cc_form() {
+		return isset( $this->settings['inline_cc_form'] ) && 'yes' === $this->settings['inline_cc_form'];
+	}
+
+	/**
+	 * Return the locale for the payment button.
+	 *
+	 * @return string Defaults to en_US.
+	 */
+	private function get_button_locale() {
+		return apply_filters( 'wc_stripe_payment_request_button_locale', substr( get_locale(), 0, 2 ) );
+	}
+
+	/**
+	 * Return the icons urls.
+	 *
+	 * @return array Arrays of icons metadata.
+	 */
+	private function get_icons() {
+		$icons_src = [
+			'visa'       => [
+				'src' => WC_STRIPE_PLUGIN_URL . '/assets/images/visa.svg',
+				'alt' => __( 'Visa', 'woo-gutenberg-products-block' ),
+			],
+			'amex'       => [
+				'src' => WC_STRIPE_PLUGIN_URL . '/assets/images/amex.svg',
+				'alt' => __( 'American Express', 'woo-gutenberg-products-block' ),
+			],
+			'mastercard' => [
+				'src' => WC_STRIPE_PLUGIN_URL . '/assets/images/mastercard.svg',
+				'alt' => __( 'Mastercard', 'woo-gutenberg-products-block' ),
+			],
+		];
+
+		if ( 'USD' === get_woocommerce_currency() ) {
+			$icons_src['discover'] = [
+				'src' => WC_STRIPE_PLUGIN_URL . '/assets/images/discover.svg',
+				'alt' => __( 'Discover', 'woo-gutenberg-products-block' ),
+			];
+			$icons_src['jcb']      = [
+				'src' => WC_STRIPE_PLUGIN_URL . '/assets/images/jcb.svg',
+				'alt' => __( 'JCB', 'woo-gutenberg-products-block' ),
+			];
+			$icons_src['diners']   = [
+				'src' => WC_STRIPE_PLUGIN_URL . '/assets/images/diners.svg',
+				'alt' => __( 'Diners', 'woo-gutenberg-products-block' ),
+			];
+		}
+		return $icons_src;
```

Comment:
We need to somehow [apply the `woocommerce_gateway_icon` filter](https://github.com/woocommerce/woocommerce-gateway-stripe/blob/0c182f1dc17614207c5c559abcae220b9735882e/includes/class-wc-gateway-stripe.php#L206) here.

The non-block parts of the plugin provide a string to this filter that has a **very different structure**, so that's going to be difficult, but I think we need to provide parity here, and allow merchants to filter the accepted payment method icons.

We might have to introduce an entirely new filter for this, or somehow extract data out of a big string [full of `<img />` elements](https://github.com/woocommerce/woocommerce-gateway-stripe/blob/0c182f1dc17614207c5c559abcae220b9735882e/includes/abstracts/abstract-wc-stripe-payment-gateway.php#L212-L232).

It's potentially enough to just apply the [`wc_stripe_payment_icons` filter](https://github.com/woocommerce/woocommerce-gateway-stripe/blob/0c182f1dc17614207c5c559abcae220b9735882e/includes/abstracts/abstract-wc-stripe-payment-gateway.php#L212-L213) here, and maybe we should be applying the `woocommerce_gateway_icon` filter [in the frontend](https://github.com/woocommerce/woocommerce-gateway-stripe/blob/0c182f1dc17614207c5c559abcae220b9735882e/client/blocks/credit-card/payment-method.js#L20-L30)?

Actually, @nerrad does the blocks extension apply the `woocommerce_gateway_icon` filter for us?

Reply:
Blocks doesn't touch this at all, the Stripe integration code is just calling this and passing icons through to the client side registration for Stripe.

With the new integration introduced, we're moving away from a global icons type thing and having payment methods configure and provide their icons. We _intentionally_ left filters out on the blocks side here because I think that's something we may need to do a good job of planning around to:

- handle scenarios where there may be duplication of icons.
- aim for an interface that restricts what can be added via this filter (right now it's anything goes - which seems...dangerous in the context of payment methods)?
- explore different ways that payment method icons can be managed (would that be a good ui/ux option for merchants to have some control over?)
- leave room for potentially offering consistent icon style "packages/patterns" that merchants can choose from.

By adding a filter, we implement something that has to be supported indefinitely into the future. All that to say, if you think Stripe as a payment method should retain the flexibility of custom icons added by a filter (with all the risks/baggage that brings) in the integration, that's a team for your team to decide and is outside the purview of what the Blocks API _currently_ handles.

Reply:
> if you think Stripe as a payment method should retain the flexibility of custom icons added by a filter in the integration, that's something for your team to decide and is outside the purview of what the Blocks API currently handles.

Ok, good to know! Thank you for the clarification!

I do think we need to retain this for compatibility with the shortcode checkout, but I share your concerns — we'll have to make sure everything is properly escaped.

That said, these concerns apply to most filters, so that's nothing terribly new 🙂

Reply:
Important to keep in mind, that the new checkout flow does _not_ have the goal of parity with the previous shortcode flow in terms of compatibility and design. I realize you may be referring to compatibility in terms of sharing code between the flows, but I do encourage avoiding simply adding filters here because the shortcode flow has them. There's opportunity to think through _why_ these filters are in place and whether the problem being solved can be solved in a way that preserves some of the direction being taken with the new flows.

Reply:
That's an excellent point @nerrad! I made a note of this in the relevant issue: https://github.com/woocommerce/woocommerce-gateway-stripe/issues/1517#issuecomment-828000648 🙂

Reply:
Just to be clear; this will be solved with #1517 . I'm not marking this discussion as resolved, just for visibility.